src/kvp.rs

```diff
@@ -494,6 +563,76 @@ fn get_uptime() -> Duration {
     Duration::from_secs(uptime_seconds)
 }
 
+/// Represents the type of health report for provisioning.
+#[derive(PartialEq)]
+enum HealthReportKind {
+    Success,
+    Failure,
+}
+
+impl HealthReportKind {
+    pub fn from_str(s: &str) -> Self {
+        match s {
+            "failure" => HealthReportKind::Failure,
+            "success" => HealthReportKind::Success,
+            _ => panic!("Invalid HealthReportKind: {}", s),
+        }
+    }
+}
+
+/// Builds the provisioning health KVP entry in the required format.
+///
+/// For a failure, the expected format is:
+///   result=error|reason=<failure reason>|agent=Azure-Init/<version>|extra=stuff|pps_type=None|vm_id=<vm_id>|timestamp=<timestamp>|documentation_url=https://aka.ms/linuxprovisioningerror
+///
+/// For success, the expected format is:
+///   result=success|agent=Azure-Init/<version>|pps_type=None|vm_id=<vm_id>|timestamp=<timestamp>[|optional_key=optional_value…]
+///
+/// # Parameters
+/// - `status`: Specifies whether this is a success or failure report.
+/// - `reason`: For failure reports, the failure reason. Must be provided for failures.
+/// - `extra`: Optional additional key=value pairs.
+/// - `vm_id`: The VM ID.
+/// - `timestamp`: The precomputed timestamp in ISO 8601 format.
+///
+/// # Returns
+/// A tuple of (key, value) where key is fixed as "PROVISIONING_REPORT".
+fn build_health_kvp(
```

Comment:
I think it would be best to split this into two interfaces.

Avoid the need for HealthReportKind and split into report_success_kvp() / report_failure_kvp().  Is there value in keeping these together?

Reply:
while using types, this is very much bool like.  if the boolean isn't trivial change in a moethod, I would consider making two methods which can each do their own thing with correct interfaces.  easier to read, and easier to maintain over time in my opinion.